src/Microsoft.ML.Dnn/TensorTypeExtensions.cs

```diff
@@ -0,0 +1,67 @@
+﻿using System;
+using NumSharp.Backends;
+using NumSharp.Backends.Unmanaged;
+using NumSharp.Utilities;
+using Tensorflow;
+
+namespace Microsoft.ML.Transforms
+{
+    [BestFriend]
+    internal static class TensorTypeExtensions
+    {
+        public static void ToScalar<T>(this Tensor tensor, ref T dst) where T : unmanaged
+        {
+            if (typeof(T).as_dtype() == tensor.dtype)
+            {
+                unsafe
+                {
+                    dst = *(T*)tensor.buffer;
+                }
+            }
+        }
+
+        public static void ToSpan<T>(this Tensor tensor, Span<T> values) where T: unmanaged
+        {
+            if (typeof(T).as_dtype() == tensor.dtype)
+            {
+                unsafe
+                {
+                    var len = (long)tensor.size;
+                    fixed (T* dst = values)
+                    {
+                        var src = (T*)tensor.buffer;
+                        len *= ((long)tensor.itemsize);
+                        System.Buffer.MemoryCopy(src, dst, len, len);
```

Comment:
Doing it this way is dangerous because we don't know if the `values` span is big enough. So copying a too big buffer into it will trash memory.

Instead, I would write this the other way. Turn the `tensor.buffer` into a `Span<T>`, and then use `Span.CopyTo(Span<T> destination)` to do the copy. That will throw if there isn't enough space.

Note:

The way you create a `Span<T>` is to use this constructor:

```
        public Span(void* pointer, int length);
```

Reply:
When I use CopyTo method I get a test failure:
InvalidOperationException : Splitter/consolidator worker encountered exception while consuming source data
    ---- System.ArgumentException : Destination is too short.

Reply:
That's perfect! You caught a bug. You should see why your destination is too short.